usr/share/lib/ipa/tests/conftest.py

```diff
@@ -39,6 +34,43 @@ def f(repo):
     return f
 
 
+@pytest.fixture()
+def determine_provider(host):
+    def f():
+        result = host.run('sudo dmidecode -t system')
+        output = result.stdout.lower()
+        if 'amazon' in output:
+            provider = 'ec2'
+        elif 'microsoft' in output:
+            provider = 'azure'
+        elif 'google' in output:
+            provider = 'gce'
+        else:
+            raise Exception('Provider not found.')
+        return provider
+    return f
+
+
+@pytest.fixture()
+def determine_region(host):
+    def f(provider):
+        if provider == 'ec2':
+            result = host.run('ec2metadata --availability-zone')
+            region = result.stdout.strip()[:-1]
+        elif provider == 'gce':
+            result = host.run('gcemetadata --query instance --zone')
+            region = result.stdout.strip().rsplit('/', maxsplit=1)[-1]
+        elif provider == 'azure':
+            result = host.run(
+                'curl -H Metadata:true '
+                '"http://169.254.169.254/metadata/instance'
+                '?api-version=2017-12-01"'
+            )
```

Comment:
Use python-requests or urllib3 instead of running an external process

Reply:
pytest is running on the tester system whereas the command is run on the host (testee).

E.g. the only option to use python is invoke as command:

`python -c "import requests;r = requests.get('http://169.254.169.254/metadata/instance?api-version=2017-12-01', headers={'Metadata': 'true'}...etc)"`

But this won't work if we want to use requests since we don't want to force install a new package in the middle of testing an instance.

I think curl is the best option.